OrbitVulkanLayer/SubmissionTracker.h

```diff
@@ -0,0 +1,757 @@
+  std::string text;
+  Color color;
+  size_t depth;
+  bool cut_off;
+};
+
+// A single submission (VkQueueSubmit) can contain multiple `SubmitInfo
+struct SubmitInfo {
+  std::vector<SubmittedCommandBuffer> command_buffers;
+};
+
+// Wraps up all the data that needs to be persistent upon a submission (VkQueueSubmit).
+// `completed_markers` are all the debug markers, that got completed (via "End") within this
+// submission. Their "Begin" might still be in a different submission.
+struct QueueSubmission {
+  SubmissionMetaInformation meta_information;
+  std::vector<SubmitInfo> submit_infos;
+  std::vector<MarkerState> completed_markers;
+  uint32_t num_begin_markers = 0;
+};
+
+}  // namespace internal
+
+/*
+ * This class ultimately is responsible to track command buffer and debug marker timings.
+ * To do so, it keeps tracks of command-buffer allocations, destruction, begins, ends as well as
+ * submissions.
+ * On `VkBeginCommandBuffer` and `VkEndCommandBuffer` it can (if capturing) insert write timestamp
+ * commands (`VkCmdWriteTimestamp`). The same is done for debug marker begins and ends. All that
+ * data will be gathered together at a queue submission (`VkQueueSubmit`).
+ *
+ * Upon every `VkQueuePresentKHR` it will check if the timestamps of a certain submission are
+ * already available, and if so, it will send the results over to the `VulkanLayerProducer`.
+ *
+ * See also `DispatchTable` (for vulkan dispatch), `TimerQueryPool` (to manage the timestamp slots),
+ * and `DeviceManager` (to retrieve device properties).
+ *
+ * Thread-Safety: This class is internally synchronized (using read/write locks), and can be
+ * safely accessed from different threads.
+ */
+template <class DispatchTable, class DeviceManager, class TimerQueryPool>
+class SubmissionTracker : public VulkanLayerProducer::CaptureStatusListener {
+ public:
+  explicit SubmissionTracker(uint32_t max_local_marker_depth_per_command_buffer,
+                             DispatchTable* dispatch_table, TimerQueryPool* timer_query_pool,
+                             DeviceManager* device_manager)
+      : max_local_marker_depth_per_command_buffer_(max_local_marker_depth_per_command_buffer),
+        dispatch_table_(dispatch_table),
+        timer_query_pool_(timer_query_pool),
+        device_manager_(device_manager) {
+    CHECK(dispatch_table_ != nullptr);
+    CHECK(timer_query_pool_ != nullptr);
+    CHECK(device_manager_ != nullptr);
+  }
+
+  // Sets a producer to be used to enqueue capture events and ask if we are capturing.
+  // We will also register our self as CaptureStatusListener, in order to get nofified on
+  // capture finish (OnCaptureFinished). Will reset the open query slots then.
+  void SetVulkanLayerProducer(VulkanLayerProducer* vulkan_layer_producer) {
+    vulkan_layer_producer_ = vulkan_layer_producer;
+    if (vulkan_layer_producer_ != nullptr) {
+      vulkan_layer_producer_->SetCaptureStatusListener(this);
+    }
```

Comment:
I thought about that and maybe this is cleaner. But on the other hand, when we call `SetVulkanLayerProducer` with nullptr, the producer will get destroyed, and it would not matter if we are still register. But there would be the risk, that the pointer to the "old" producer is already invalid. This is why I decided to omit this. But I am open to add the call here. WDYT?

Reply:
Yeah, if the client of the class calls `SetVulkanLayerProducer(nullptr)` after having destroyed the `VulkanLayerProducer`, that would be a problem? But isn't it wrong? The client shouldn't destroy an object someone has a pointer to, as we could still be making calls to it.
On the other hand, if we don't de-register it could in theory happen that the client calls `SetVulkanLayerProducer(nullptr)` but the `VulkanLayerProducer` still exists and we are still receiving notifications from it.

Reply:
>Yeah, if the client of the class calls SetVulkanLayerProducer(nullptr) after having destroyed the VulkanLayerProducer, that would be a problem?

Well then the call to unset the listener would fail, right?

>The client shouldn't destroy an object someone has a pointer to, as we could still be making calls to it.

Agree, but there is still the risk that this might happen e.g. based on some refactoring in the future.

>On the other hand, if we don't de-register it could in theory happen that the client calls SetVulkanLayerProducer(nullptr) but the VulkanLayerProducer still exists and we are still receiving notifications from it.

Right, so the question is, what is more likely and what would cause more problems.
For me both solutions seem kind of fine. Which one would you choose?